adobe/copy_on_write.hpp

```diff
@@ -55,93 +36,111 @@ copy when the value is modified and there is more than one reference to the
 
     /*!
-        Initializes the instance with a custom allocator
+    \brief Constructs a new copy_on_write object with a single argument.
     */
-    explicit copy_on_write(const allocator_type& a) : object_m(0) {
-        implementation_allocator_type other_allocator(a);
-
-        object_m = allocate(other_allocator);
-    }
+    template <class U>
+    copy_on_write(U&& x, disable_copy<U> = nullptr) : _self(new model(std::forward<U>(x))) {}
 
     /*!
-    \brief Constructs a new copy_on_write object with a value \c x.
-
-    \param x A default value to assign to this object
+    \brief Constructs a new copy_on_write object with two or more arguments.
     */
-    copy_on_write(T x) : object_m(allocate_move(0, std::move(x))) {}
+    template <class U, class V, class... Args>
+    copy_on_write(U&& x, V&& y, Args&&... args)
+        : _self(new model(std::forward<U>(x), std::forward<V>(y), std::forward<Args>(args)...)) {}
 
     /*!
     Copy construction is a non-throwing operation and simply increments the
     reference count on the stored object.
     */
-    copy_on_write(const copy_on_write& x) : object_m(x.object_m) {
-        if (object_m)
-            object_m->header_m.get().count_m.increment();
-    }
+    copy_on_write(const copy_on_write& x) noexcept : _self(x._self) {
+        assert(_self && "FATAL (sparent) : using a moved copy_on_write object");
 
-    copy_on_write(copy_on_write&& x) : object_m(x.object_m) { x.object_m = 0; }
+        ++_self->_count;
+    }
+    copy_on_write(copy_on_write&& x) noexcept : _self(x._self) {
+        assert(_self && "WARNING (sparent) : using a moved copy_on_write object");
+        x._self = nullptr;
+    }
 
-    ~copy_on_write() { release(object_m); }
+    ~copy_on_write() { destroy(); }
 
     /*!
         As with copy construction, assignment is a non-throwing operation which
         releases the old value and increments the reference count of the item
         being assigned to.
     */
-    copy_on_write& operator=(copy_on_write x) {
-        swap(*this, x);
-        return *this;
+    auto operator=(const copy_on_write& x) noexcept -> copy_on_write& {
+        return *this = copy_on_write(x);
     }
+    auto operator=(copy_on_write&& x) noexcept -> copy_on_write& {
+        assert(x._self && "WARNING (sparent) : using a moved copy_on_write object");
+        assert((&x != this) && "FATAL (sparent) : invalid self-move.");
 
+        destroy();
+        _self = x._self;
+        x._self = nullptr;
+        return *this;
+    }
 
-    copy_on_write& operator=(T x) {
-        if (!object_m)
-            object_m = allocate_move(0, std::move(x));
-        else if (object_m->header_m.get().count_m.is_one())
-            object_m->value_m = std::move(x);
-        else
-            reset(allocate_move(object_m, std::move(x)));
+    template <class U>
+    auto operator=(U&& x) -> disable_copy_assign<U> {
+        if (_self && unique()) {
+            _self->_value = std::forward<U>(x);
```

Comment:
Yes. As thread safe as an int. For another thread to increment the ref-count that thread would have to have a reference, not a copy, to the object. No different then if I assigned to an int while another thread was reading a ref to the int.

One thing to be aware of, you can't implement cow on top of shared_ptr as of C++17. For C++11 and C++14 you could do it using shared_ptr and unique (by the standard). However several implementations were using memory_order_relaxed for their atomic count for performance. The problem with that is then you have no memory ordering guarantees so a thread could decrement the ref count but still be reading from the block because the reads are re-ordered after the decrement. C++17 it was decided that the performance was more important than unique() and unique() is now deprecated. I'm currently using the default memory_order_seq_cst - which guarantees it will work. If I spent enough time contemplating it I might be able to relax to memory_order_aquire (on increment) and memory_order_release (on decrement) but it wouldn't matter for intel and I'd want to pull in another expert or two to review if I got that aggressive.